### DISCUSSION TITLE
Dynamic Imports SSR and client side routing

### DISCUSSION THREAD

**Discussion by AlexPagnotta** [vercel/next.js#32923] [Help]
Hi!

I am building a static site using NextJS (SSG only) and contentful as CMS, I have a single page in my project, whose components are dynamically rendered based on the data returned from contentful in the getStaticProps function.
So on my page, I need to have a big "switch" statement to conditionally render the appropriate components, something like this:

```
import React from 'react'

import Component1 from './components/Component1'
import Component2 from './components/Component2'
import Component3 from './components/Component3'
import Component4 from './components/Component4'

type Props = {
  data: HeaderComponentFragment
}

const Switcher = ({ data }: Props): JSX.Element => {
  return (
    <>
      {data.componentName === 'Component1' ? (
        <Component1 {...data} />
      ) : data.componentName === 'Component2' ? (
        <Component2 {...data} />
      ) : data.componentName === 'Component3' ? (
        <Component3 {...data} />
      ) : data.componentName === 'Component4' ? (
        <Component4 {...data} />
      )  : null}
    </>
  )
}

export default Switcher
```

I'm importing the components dynamically, as without it I would need to import all the components on every page, even the one that I am not using.

The only problem is that when I switch page client, using a next Link, the components are not server side rendered anymore, so I see a lot of layout shifts cause all the components are loading, and the experience is not optimal.

I am not sure if this is the expected behaviour, a dynamic imports limitation, or an error on my side, but I suppose this is a common use case, so maybe some of you have a solution.

Thanks for the help!

**Comment by TommySorensen** [comment]
You basically want to to the same like this solution where every single component is bundled in it's own file:
https://github.com/vercel/next.js/discussions/11291?sort_order=chronological#discussioncomment-4303

**Reply by AlexPagnotta** [reply]
Yes, it's the same problem, more or less, but I already implement the suggested solution, dynamic imports with ssr: true, and It works fine, SSR works too, no layout shifts when the page is loaded.
The problem appears when the page Is reached using a next link, so a client side route, in this case there is no html returned by the server to act as a "placeholder", and the components are loaded dynamically, so I see a lot of layout shifts.

What I am trying to understand is if this is the normal behaviour, and if there is something I can do about it, the only things that come to my mind is to set a spinner while waiting for the component to load (not the best solution),  or to dynamically load only the components below the fold.
